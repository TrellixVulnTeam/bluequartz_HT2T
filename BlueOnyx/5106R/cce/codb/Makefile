# $Id: Makefile 259 2004-01-03 06:28:40Z shibuya $
# Copyright 2001 Sun Microsystems, Inc.  All rights reserved.

include ../dynamic.mk

DIRLIB = codb.a
OBJS = src/codb.o

all: $(DIRLIB) 

$(DIRLIB): $(OBJS)
	$(AR) rcs $(DIRLIB) $(OBJS)

$(OBJS): FORCE
	$(MAKE) -C src all

debug:
	$(MAKE) all DEBUG="$(CCE_DEBUG) -DDEBUG_CODB"

depend:
	$(MAKE) -C src depend
	
test:
	$(MAKE) -C test

clean:
	$(MAKE) -C src clean
	$(MAKE) -C test clean
	$(RM) -f $(DIRLIB)

FORCE:

# dependency info
ifeq (.depend,$(wildcard .depend))
include .depend
endif
