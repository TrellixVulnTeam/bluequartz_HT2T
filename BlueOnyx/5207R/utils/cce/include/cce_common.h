/* $Id: cce_common.h 623 2005-10-30 15:53:49Z shibuya $ */
/* Copyright 2001 Sun Microsystems, Inc.  All rights reserved. */

#ifndef __CCE_COMMON_H__
#define __CCE_COMMON_H__

#include <stdlib.h>
#include <stdio.h>
#include <errno.h>
#include <syslog.h>
#include <time.h>
#include <sys/time.h>
#include <sys/types.h>
#include <unistd.h>

#include <cce_paths.h>

/*****
 * flags associated with debug print/syslog/error printing.
 *****/

/* the global debug mask - 0 = no debugging */
extern unsigned long cce_debug_mask;

/* the 'verbose' flag.  send syslog's to stderr */
extern int vflag;

/* the 'no logging' flag.  never send anything to syslog */
extern int nologflag;

/* "fail everything" flag. */
extern int txnfailflag;

/* don't run transaction support library */
extern int txnstopflag;

/*****
 * some macros to manipulate them
 *****/
#define set_debug_mask(m)	do { cce_debug_mask = (m); } while (0)
#define add_debug_mask(m)	do { cce_debug_mask |= (m); } while (0)
#define rm_debug_mask(m)	do { cce_debug_mask &= ~(m); } while (0)


/*****
 * the functions/macros to output stuff
 *****/
/*
 * cce_syslog
 * 	if (!nl) syslog
 * 	if (-V) stderr
 *
 * DPRINTF
 * 	if (-d <DBG_X>) stderr
 */

/* 
 * use DPRINTF()/DPERROR() when there is data to print, but it is not
 * fatal.  We can then enable the debugging flag for each subsystem to get
 * more info on it.
 */
#define DPRINTF(m, ...) 	\
	do { \
		struct timeval ts; \
		struct tm *tm; \
		gettimeofday(&ts, NULL); \
		if (cce_debug_mask & (m)) { \
			fprintf(stderr, "%02d:%02d:%02d.%ld [%li] (%s:%d): ", \
				ts.tv_usec , \
				(long)getpid() , __FILE__ , __LINE__); \
			fprintf(stderr, __VA_ARGS__); \
		} \
	} while (0)

#define DPRINTF_CORE(m, ...) \
	do { \
		if (cce_debug_mask & (m)) { \
			fprintf(stderr, __VA_ARGS__); \
		} \
	} while (0)

#define DPERROR(m, ...) \
	do { \
		DPRINTF(m, __VA_ARGS__); \
		DPRINTF_CORE(m, ": %s\n" , strerror(errno)); \
	} while (0)

#define CCE_SYSLOG(...) \
	do { \
		if (vflag) { \
			fprintf(stderr, __VA_ARGS__); \
			fprintf(stderr, "\n"); \
		} \
		if (!nologflag) \
			syslog(LOG_INFO, __VA_ARGS__); \
	} while (0)

#define DPROFILE_START(m, tvp, ...) \
	do { \
		if (cce_debug_mask & m) { \
			gettimeofday((tvp), NULL); \
			DPRINTF(m, "PROFILING: "); \
			DPRINTF_CORE(m, __VA_ARGS__); \
			DPRINTF_CORE(m, "\n"); \
		} \
	} while (0)	

#define DPROFILE(m, tv, ...) \
	do { \
		if (cce_debug_mask & m) { \
			struct timeval tv2; \
			long unsigned duration; \
			gettimeofday(&tv2, NULL); \
			duration = (tv2.tv_sec - (tv).tv_sec) * (1000000) \
				+ (tv2.tv_usec - (tv).tv_usec); \
			DPRINTF(m, "-- TIME: %ld.%06ld: " , \
				(long)(duration / 1000000) , \
				(long)(duration % 1000000)); \
			DPRINTF_CORE(m, __VA_ARGS__); \
			DPRINTF_CORE(m, "\n"); \
		} \
	} while (0)

/* debugging values for cce_debug_mask, used throughout cce */
#define DBG_NONE    		0x0000
#define DBG_CCED    		0x0001
#define DBG_CODB    		0x0002
#define DBG_ED      		0x0004
#define DBG_CONF    		0x0008
#define DBG_SCHEMA  		0x0010
#define DBG_SESSION 		0x0020
#define DBG_CSCP    		0x0040
#define DBG_SCALAR   		0x0080
#define DBG_COMMON   		0x0100
#define DBG_CSCP_XTRA		0x0200
#define DBG_TXN      		0x0400
#define DBG_EXCESSIVE		0x8000

/* profiling values */
#define PROF_NONE    		0x00000000
#define PROF_CCED    		0x00010000
#define PROF_CODB    		0x00020000
#define PROF_ED      		0x00040000
#define PROF_CONF    		0x00080000
#define PROF_SCHEMA  		0x00100000
#define PROF_SESSION 		0x00200000
#define PROF_CSCP    		0x00400000
#define PROF_SCALAR   		0x00800000
#define PROF_COMMON   		0x01000000
#define PROF_CSCP_XTRA		0x02000000
#define PROF_TXN      		0x04000000

#endif /* __CCE_COMMON_H__ */

#ifdef USE_LIBDEBUG
#define USE_MEMDEBUG 2
#include <libdebug.h>
#endif

/* Copyright (c) 2003 Sun Microsystems, Inc. All  Rights Reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met: 
 * 
 * -Redistribution of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 * 
 * -Redistribution in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution. 
 *
 * Neither the name of Sun Microsystems, Inc. or the names of contributors may
 * be used to endorse or promote products derived from this software without 
 * specific prior written permission.

 * This software is provided "AS IS," without a warranty of any kind. ALL EXPRESS OR IMPLIED CONDITIONS, REPRESENTATIONS AND WARRANTIES, INCLUDING ANY IMPLIED WARRANTY OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE OR NON-INFRINGEMENT, ARE HEREBY EXCLUDED. SUN MICROSYSTEMS, INC. ("SUN") AND ITS LICENSORS SHALL NOT BE LIABLE FOR ANY DAMAGES SUFFERED BY LICENSEE AS A RESULT OF USING, MODIFYING OR DISTRIBUTING THIS SOFTWARE OR ITS DERIVATIVES. IN NO EVENT WILL SUN OR ITS LICENSORS BE LIABLE FOR ANY LOST REVENUE, PROFIT OR DATA, OR FOR DIRECT, INDIRECT, SPECIAL, CONSEQUENTIAL, INCIDENTAL OR PUNITIVE DAMAGES, HOWEVER CAUSED AND REGARDLESS OF THE THEORY OF LIABILITY, ARISING OUT OF THE USE OF OR INABILITY TO USE THIS SOFTWARE, EVEN IF SUN HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.
 * 
 * You acknowledge that  this software is not designed or intended for use in the design, construction, operation or maintenance of any nuclear facility.
 */
