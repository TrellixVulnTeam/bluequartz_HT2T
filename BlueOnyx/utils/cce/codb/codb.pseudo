/* $Id: codb.pseudo 259 2004-01-03 06:28:40Z shibuya $ */
/* Copyright 2001 Sun Microsystems, Inc.  All rights reserved. */
/*
 * Tim Hockin
 *
 * This is a proposed API and some pseudocode for a new version of the
 * Cobalt Object Database.
 *
 * This version should be faster, cleaner, and more extendable 
 * Add to these ideas a stacked layer model of services to run on gets and
 * commits.
 *
 * Cache needs to be either in shared memory, or in a single process,
 * through which all object accesses go.
 *
 * We'd need to decide on whether objects are volatile or not:
 *	Time 0:	A = new handle		B = new handle
 *	Time 1:	A gets oid n		B gets oid n
 *	Time 2: 	A read n.foo as 0		B reads n.bar as 0
 *	Time 3:	A sets n.foo = 1		B sets n.bar = 1
 *	Time 4:	A commits		
 *	Time 5:	A reads n			B reads n
 *
 * Question:  does B see foo == 1 or foo == 0 ?
 *   volatile:  A gets foo == 1, bar == 0 ; B gets foo == 1, bar == 1
 *   non-volatile: A gets foo == 1, bar == 0 ; B gets foo == 0, bar == 1
 */
 

/*
 * fundamental datatypes
 */

typedef long			oid_t;		/* note signedness... */
typedef enum {...} 		codb_ret_t;
typedef struct _codb_handle 	codb_handle;	/* all these are OPAQUE */	
typedef struct _codb_obj 	codb_obj;
typedef struct _codb_props	codb_props;


/*
 * summary of exported functions
 */

/* init */
codb_ret_t 		codb_init(void);

/* handle functions */
codb_handle *	codb_handle_new(void);
void 			codb_handle_destroy(codb_handle *);
codb_ret_t 		codb_handle_errval(codb_handle *);

/* cache functions - probably only internal */
void 			codb_expire_ocache(int s);
void 			codb_flush_ocache(void);

/* object operatiosn */
char *		codb_obj_get_prop(codb_handle *, codb_obj *, "prop");
codb_obj *		codb_obj_assign(codb_obj *);
void 			codb_obj_release(codb_obj *);

/* database operations */
codb_obj *		codb_get(codb_handle *, oidt_t, "namespace");
char *		codb_get_prop(codb_handle *, oidt_t, "prop");
int 			codb_exists(codb_handle *, oid_t);
codb_ret_t 		codb_set(codb_handle *, codb_obj *, codb_props *);
codb_obj *		codb_create(codb_handle *, "class", codb_props *);
codb_ret_t 		codb_destroy(codb_handle *, codb_obj *);
GSList *		codb_find(codb_handle *, "class", codb_props *);
GSList *		codb_names(codb_handle *, "class");
GSList *		codb_classlist(codb_handle *);
codb_ret_t 		codb_commit(codb_handle *);
codb_ret_t 		codb_cancel(codb_handle *);

/* properties operations */
codb_props *	codb_props_new(void);
void 			codb_props_destroy(codb_props *);
int 			codb_props_set(codb_props *, "prop", "val");
char *		codb_props_get(codb_props *, "prop");
void 			codb_props_clear(codb_props *);
codb_props *	codb_props_merge(codb_props *, codb_props *);



/*
 * codb exported functions
 *
 * always handle NULL codb_handle * with an error
 * always set handle.errval (except accessor function for errval)
 * always handle bad params
 */

/* initialize */
codb_ret_t
codb_init(void):
{
	run all initializers
	set alarm to run codb_expire_ocache
}


/* 
 * create new codb_handle 
 * allocates memory
 */
codb_handle *
codb_handle_new(void):
{
	allocate codb_handle
	set handle.errval
	return new codb_handle
}


/* 
 * clean up a codb_handle 
 */
void
codb_handle_destroy(codb_handle *);
{
	free memory in handle
}


/* 
 * get the most recent errval from a handle 
 */
codb_ret_t
codb_handle_errval(codb_handle *):
{
	ASSERT(handle != NULL)

	return most recent errval
}


/*
 * flush old items in ocache
 */
void
codb_expire_ocache(int s):
{
	for each obj in ocache {
		if (refcount = 0 && lastused > 5 minutes ago) {
			flush_obj_from_ocache(obj);
		}
	}
}


/*
 * flush all ocache
 */
void
codb_flush_ocache(void):
{
	for each obj in ocache {
		flush_obj_from_ocache(obj);
	}
}


/* 
 * get a property from a codb_obj pointer 
 * this can be called, even after an obj is detroyed, but not commited
 */
char *
codb_obj_get_prop(codb_handle *, codb_obj *, "prop"):
{
	ASSERT(handle != NULL)

	if (!obj || !prop) {
		set handle.errval
		return NULL
	}

	set handle.errval to success;

	if (obj is in handle.mod_cache) {
		return value from handle.mod_cache
	} else {
		return value from obj
	}
}
		

/* 
 * assign the codb_obj *, while incrementing refcount
 */
codb_obj *
codb_obj_assign(codb_obj *):
{
	if (obj)
		increment obj refcount
	}
	return obj
}


/* 
 * decrement refcount
 */
void
codb_obj_release(codb_obj *):
{
	if (obj) {
		decrement obj refcount
		if refcount == 0
		set lastused to current time
	}
}


/* 
 * get a codb_obj pointer 
 */
codb_obj *
codb_get(codb_handle *, oidt_t, "namespace"):
{
	ASSERT(handle != NULL)

	if (oid <= 0) {
		set handle.errval
		return NULL
	}

	if (!namespace) {
		namespace = ""
	}

	if (oid+namespace is in handle.destroy_cache) {
		set handle.errval
		return NULL;
	}

	set handle.errval

	if (oid+namespace is in handle.create_cache) {
		increment refcount
		return pointer to obj in create_cache
	}

	if (oid+namespace is not in ocache) {
		load object into ocache
		if (object does not exist) {
			set handle.errval
			return NULL
		}
	}

	increment refcount
	return pointer to obj in ocache;
}


/*
 * FIXME: need a get_old and get_changed
 */


/* 
 * get a property of an oid
 */
char *
codb_get_prop(codb_handle *, oidt_t, "prop"):
{
	ASSERT(handle != NULL)

	obj = codb_get(handle, oid, NULL);
	ret = codb_obj_get_prop(handle, obj, prop);
	codb_obj_release(obj);

	return ret;
}
	
/* 
 * test that oid exists
 */
#define codb_exists(h, oid)	(codb_get_prop(h, oid, "CLASS") ? 1 : 0)


/* 
 * set a codb_props on a codb_obj 
 * copies data from props (allocates memory, to be freed internally)
 */
codb_ret_t
codb_set(codb_handle *, codb_obj *, codb_props *):
{
	ASSERT(handle != NULL)

	if (!obj) {
		set handle.errval
		return handle.errval;
	}

	set handle.errval

	if (!props) {
		return handle.errval
	}
		
	if (obj is in handle.destroy_cache) {
		set handle.errval
		return handle.errval;
	}

	if (obj is in handle.modify_cache) {
		merge changes with modify_cache
	} else {
		create changes in codb_handle
	}

	return handle.errval;
}


/* 
 * create a new codb_obj, and set props on it 
 * copies data from props (allocates mem, to be freed internally)
 */
codb_obj *
codb_create(codb_handle *, "class", codb_props *):
{
	ASSERT(handle != NULL)

	if (!class) {
		set handle.errval
		return NULL;
	}
		
	allocate new oid
	allocate new codb_obj in create_cache
	if (props) {
		call codb_set(handle, newobj, props
	}
	set handle.errval
	
	increment refcount
	return newobj
}


/* destroy a codb_obj */
codb_ret_t
codb_destroy(codb_handle *, codb_obj *):
{
	ASSERT(handle != NULL)

	if (!obj) {
		set handle.errval
		return handle.errval;
	}

	if (obj is not in handle.destroy_cache) {
		if (obj is in handle.create_cache) {
			decrement refcount
			remove from handle.create_cache;
		}
		increment refcount
		put codb_obj in handle.destroy_cache
		set handle.errval
	} else {
		set handle.errval
	}

	return handle.errval
}


/*
 * get a list (of codb_obj *) of objects of specified class that match
 * criteria
 */
GSList *
codb_find(codb_handle *, "class", codb_props *):
{
	ASSERT(handle != NULL)

	if (!class) {
		/* find all objects */
		class = "";
	}

	look at index for class
	for each item in index {
		if (!props) {
			increment refcount
			add to list
			continue
		}
		for each property in criteria {
			get value (through codb_handle)
			compare
			if (equal) {
				increment refcount
				add to list
			}
		}
	}
	set handle.errval
	return list
}


/* 
 * get a list (of strings) of all namespaces of a class
 * strings are static, do not free 
 */
GSList *
codb_names(codb_handle *, "class"):
{
	ASSERT(handle != NULL)

	if (!class) {
		set handle.errval
		return NULL
	}

	query classconf layer
	set handle.errval
	return list
}


/* 
 * get a list (of strings) of all classes on the system 
 * strings are static, do not free 
 */
GSList *
codb_classlist(codb_handle *):
{
	ASSERT(handle != NULL)

	query classconf layer
	set handle.errval
	return list
}


/* sync codb_handle buffers with database */
codb_ret_t 
codb_commit(codb_handle *):
{
	ASSERT(handle != NULL)

	create objects in handle.create_cache
	modify objects in handle.mod_cache
	destroy objects in handle.destroy_cache

	set handle.errval
	return list
}


/* flush pending changes */
codb_ret_t 
codb_cancel(codb_handle *):
{
	ASSERT(handle != NULL)

	clear handle.create_cache
	clear handle.mod_cache
	clear handle.destroy_cache

	set handle.errval
	return list
}




/*
 * codb_props
 *
 * none of these functions allocate memory for keys or values
 * they use the pointers AS PASSED
 */

/* new codb_props */
codb_props *
codb_props_new(void):
{
	return a new GHashTable
}


/* destroy a codb_props */
void 
codb_props_destroy(codb_props *):
{
	if (props) {
		call codb_props_clear(props)
		destroy the GHashTable
	}
}


/* set a property in a codb_props (return set value or NULL) */
int
codb_props_set(codb_props *, "prop", "val"):
{
	if (!props || !prop) {
		return -1
	}

	ret = hash_lookup(prop)
	if (ret) {
		remove prop from hash
	}

	set prop

	return 0
}


/* get a property from a codb_props */
char *
codb_props_get(codb_props *, "prop"):
{
	if (prop) {
		return hash_lookup(prop)
	} else {
		return NULL
	}
}


/* clear all properties in a codb_props */
void 
codb_props_clear(codb_props *):
{
	if (prop) {
		for each item in hash {
			remove item from hash
		}
	}
}


/* 
 * merge a codb_props into a new codb_props 
 * if both hashes have same key, value in first hash wins
 * allocates memory for a new codb_props structure
 */
codb_props *
codb_props_merge(codb_props *, codb_props *):
{
	create new hash
	if (props1) {
		for each item in props1 {
			add to newhash
		}
	}
	if (props2) {
		for each item in props2 {
			if item does not exist in newhash {
				add item to newhash
			}
		}
	}

	return newhash
}
