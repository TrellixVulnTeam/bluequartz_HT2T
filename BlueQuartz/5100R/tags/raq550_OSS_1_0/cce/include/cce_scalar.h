/* $Id: cce_scalar.h 259 2004-01-03 06:28:40Z shibuya $ */
/* Copyright 2001 Sun Microsystems, Inc.  All rights reserved. */

#ifndef _CCE_CCE_SCALAR_H_
#define _CCE_CCE_SCALAR_H_ 1

#include <stdio.h>

/* 
 * cce_scalar
 *
 * this is how scalar data is fed to / retreived from the ODB 
 */
typedef struct {
	int length; /* length (in bytes) of data pointed to */
	void *data; 
} cce_scalar;


/* macros */
#define cce_scalar_isdefined(sc)         (sc && sc->data)
#define cce_scalar_string(sc)		 ((char*)(((cce_scalar*)sc)->data))


/* constructors */
cce_scalar *cce_scalar_new_undef(void);
cce_scalar *cce_scalar_new(int size);
cce_scalar *cce_scalar_new_from_str(const char *str);
cce_scalar *cce_scalar_new_from_qstr(const char *str);
cce_scalar *cce_scalar_new_from_binstr(const char *str);
cce_scalar *cce_scalar_new_from_bin(const char *bindata, int len);
cce_scalar *cce_scalar_new_from_any(const char *str);
cce_scalar *cce_scalar_dup(cce_scalar *orig);

/* other manipulators */
cce_scalar *cce_scalar_resize(cce_scalar *s, int size);
void cce_scalar_assign(cce_scalar *l, cce_scalar *r);
void cce_scalar_undefine(cce_scalar *s);
void cce_scalar_destroy(cce_scalar *s);
#define cce_scalar_free(s) cce_scalar_destroy(s)

/* compare */
int cce_scalar_compare(cce_scalar *s1, cce_scalar *s2);

/* export */
char *cce_scalar_to_binstr(cce_scalar *s);
char *cce_scalar_to_str(cce_scalar *s);

/* file access */
cce_scalar *cce_scalar_new_from_file(char *filename);
int cce_scalar_from_file(cce_scalar *s, char *filename);
int cce_scalar_to_file(cce_scalar *s , char *filename);

#endif /* cce/cce_scalar.h */

/* Copyright (c) 2003 Sun Microsystems, Inc. All  Rights Reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met: 
 * 
 * -Redistribution of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 * 
 * -Redistribution in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution. 
 *
 * Neither the name of Sun Microsystems, Inc. or the names of contributors may
 * be used to endorse or promote products derived from this software without 
 * specific prior written permission.

 * This software is provided "AS IS," without a warranty of any kind. ALL EXPRESS OR IMPLIED CONDITIONS, REPRESENTATIONS AND WARRANTIES, INCLUDING ANY IMPLIED WARRANTY OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE OR NON-INFRINGEMENT, ARE HEREBY EXCLUDED. SUN MICROSYSTEMS, INC. ("SUN") AND ITS LICENSORS SHALL NOT BE LIABLE FOR ANY DAMAGES SUFFERED BY LICENSEE AS A RESULT OF USING, MODIFYING OR DISTRIBUTING THIS SOFTWARE OR ITS DERIVATIVES. IN NO EVENT WILL SUN OR ITS LICENSORS BE LIABLE FOR ANY LOST REVENUE, PROFIT OR DATA, OR FOR DIRECT, INDIRECT, SPECIAL, CONSEQUENTIAL, INCIDENTAL OR PUNITIVE DAMAGES, HOWEVER CAUSED AND REGARDLESS OF THE THEORY OF LIABILITY, ARISING OUT OF THE USE OF OR INABILITY TO USE THIS SOFTWARE, EVEN IF SUN HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.
 * 
 * You acknowledge that  this software is not designed or intended for use in the design, construction, operation or maintenance of any nuclear facility.
 */
