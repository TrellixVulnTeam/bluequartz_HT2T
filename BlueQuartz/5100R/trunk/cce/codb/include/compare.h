/* $Id: compare.h,v 1.5 2001/08/10 22:23:10 mpashniak Exp $ */
/* Copyright 2001 Sun Microsystems, Inc.  All rights reserved. */
/*
 *
 * declare various compare functions for use when sorting.
 *
 */

#ifndef _COMPARES_H_
#define _COMPARES_H_

#include <glib.h>

typedef struct _sortstruct {
	oid_t oid;
	cce_scalar *value;
	GSList *args;
} sortstruct;

sortfunc old_ascii_compare;
sortfunc locale_compare;
sortfunc old_numeric_compare;
sortfunc ip_compare;
sortfunc hostname_compare;

#endif

/* eof */
/* Copyright (c) 2003 Sun Microsystems, Inc. All  Rights Reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met: 
 * 
 * -Redistribution of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 * 
 * -Redistribution in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution. 
 *
 * Neither the name of Sun Microsystems, Inc. or the names of contributors may
 * be used to endorse or promote products derived from this software without 
 * specific prior written permission.

 * This software is provided "AS IS," without a warranty of any kind. ALL EXPRESS OR IMPLIED CONDITIONS, REPRESENTATIONS AND WARRANTIES, INCLUDING ANY IMPLIED WARRANTY OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE OR NON-INFRINGEMENT, ARE HEREBY EXCLUDED. SUN MICROSYSTEMS, INC. ("SUN") AND ITS LICENSORS SHALL NOT BE LIABLE FOR ANY DAMAGES SUFFERED BY LICENSEE AS A RESULT OF USING, MODIFYING OR DISTRIBUTING THIS SOFTWARE OR ITS DERIVATIVES. IN NO EVENT WILL SUN OR ITS LICENSORS BE LIABLE FOR ANY LOST REVENUE, PROFIT OR DATA, OR FOR DIRECT, INDIRECT, SPECIAL, CONSEQUENTIAL, INCIDENTAL OR PUNITIVE DAMAGES, HOWEVER CAUSED AND REGARDLESS OF THE THEORY OF LIABILITY, ARISING OUT OF THE USE OF OR INABILITY TO USE THIS SOFTWARE, EVEN IF SUN HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.
 * 
 * You acknowledge that  this software is not designed or intended for use in the design, construction, operation or maintenance of any nuclear facility.
 */
