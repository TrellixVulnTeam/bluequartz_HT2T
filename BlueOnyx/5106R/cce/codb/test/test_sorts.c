/* $Id: test_sorts.c,v 1.3 2001/08/10 22:23:13 mpashniak Exp $ */
/* Copyright 2001 Sun Microsystems, Inc.  All rights reserved. */
/* test the compare functions */

#include <cce_common.h>
#include <codb.h>
#include <cce_scalar.h>
#include <codb_classconf.h>
#include <compare.h>
#include <stdio.h>

void
test(sortfunc * func, GSList * args, const char *a, const char *b, int expect)
{
	sortstruct sa, sb;
	gint cmp0, cmp1;

	sa.args = args;
	sb.args = args;
	sa.value = cce_scalar_new_from_str(a);
	sb.value = cce_scalar_new_from_str(b);

	cmp0 = (*func) (&sa, &sb);
	cmp1 = (*func) (&sb, &sa);
	cmp1 *= -1;

	printf("  %s %c %s:\t",
	    a, (cmp0 < 0) ? '<' : ((cmp0 == 0) ? '=' : '>'), b);

	if (cmp0 != expect) {
		printf("a-b-fail ");
	}
	if (cmp1 != expect) {
		printf("b-a-fail ");
	}
	if (cmp0 == cmp1 && cmp1 == expect) {
		printf("ok.");
	}
	printf("\n");

	cce_scalar_destroy(sa.value);
	cce_scalar_destroy(sb.value);
}

int
main()
{
	test(old_numeric_compare, NULL, "0", "0", 0);
	test(old_numeric_compare, NULL, "1234", "1234", 0);
	test(old_numeric_compare, NULL, "1234", "1235", -1);
	test(old_numeric_compare, NULL, "1235", "12345", -1);
	test(old_numeric_compare, NULL, "4", "40000", -1);
	test(old_numeric_compare, NULL, "", "", 0);
	test(old_numeric_compare, NULL, "1", "", 1);
	test(old_numeric_compare, NULL, "0", "1", -1);

	// FIXME: compare currently things "0" > "", when they really
	// ought to be equal.  But, since this is just being used for
	// a sort, we can probably get away with this for now.
	//test(old_numeric_compare, NULL, "0","", 0);

	test(old_numeric_compare, NULL, "0.2", "0.15", 1);
	test(old_numeric_compare, NULL, "v0.2", "v0.15", -1);
	test(old_numeric_compare, NULL, "0.200", "0.15", 1);
	test(old_numeric_compare, NULL, "v0.200", "v0.15", 1);
	test(old_numeric_compare, NULL, "v0.200", "v0.15000", -1);

	test(old_numeric_compare, NULL, "abc", "Def", 0);
	test(old_numeric_compare, NULL, "vabc", "vDef", 0);

	test(ip_compare, NULL, "1.2.3.4", "1.2.3.4", 0);
	test(ip_compare, NULL, "1.2.3.4", "1.2.3.5", -1);
	test(ip_compare, NULL, "1.2.3.4", "1.2.4.4", -1);
	test(ip_compare, NULL, "1.2.3.4", "1.3.3.4", -1);
	test(ip_compare, NULL, "1.2.3.4", "2.2.3.4", -1);

	test(hostname_compare, NULL, "foo.com", "foo.com", 0);
	test(hostname_compare, NULL, "foo.com", "bar.com", 1);
	test(hostname_compare, NULL, "foobar.com", "bar.com", 1);
	test(hostname_compare, NULL, "barfoo.com", "bar.com", 1);
	test(hostname_compare, NULL, "foo.bar.com", "bar.bar.com", 1);
	test(hostname_compare, NULL, "foo.bar.x.com", "bar.x.com", 1);
	test(hostname_compare, NULL, "bar.foo.x.com", "foo.x.com", 1);
	test(hostname_compare, NULL, "foo.x.com", "foo.bar.x.com", 1);

	return 0;
}
/* Copyright (c) 2003 Sun Microsystems, Inc. All  Rights Reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met: 
 * 
 * -Redistribution of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 * 
 * -Redistribution in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution. 
 *
 * Neither the name of Sun Microsystems, Inc. or the names of contributors may
 * be used to endorse or promote products derived from this software without 
 * specific prior written permission.

 * This software is provided "AS IS," without a warranty of any kind. ALL EXPRESS OR IMPLIED CONDITIONS, REPRESENTATIONS AND WARRANTIES, INCLUDING ANY IMPLIED WARRANTY OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE OR NON-INFRINGEMENT, ARE HEREBY EXCLUDED. SUN MICROSYSTEMS, INC. ("SUN") AND ITS LICENSORS SHALL NOT BE LIABLE FOR ANY DAMAGES SUFFERED BY LICENSEE AS A RESULT OF USING, MODIFYING OR DISTRIBUTING THIS SOFTWARE OR ITS DERIVATIVES. IN NO EVENT WILL SUN OR ITS LICENSORS BE LIABLE FOR ANY LOST REVENUE, PROFIT OR DATA, OR FOR DIRECT, INDIRECT, SPECIAL, CONSEQUENTIAL, INCIDENTAL OR PUNITIVE DAMAGES, HOWEVER CAUSED AND REGARDLESS OF THE THEORY OF LIABILITY, ARISING OUT OF THE USE OF OR INABILITY TO USE THIS SOFTWARE, EVEN IF SUN HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH DAMAGES.
 * 
 * You acknowledge that  this software is not designed or intended for use in the design, construction, operation or maintenance of any nuclear facility.
 */
