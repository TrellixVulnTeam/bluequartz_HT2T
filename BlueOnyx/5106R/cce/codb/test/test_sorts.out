  0 = 0:	ok.
  1234 = 1234:	ok.
  1234 < 1235:	ok.
  1235 < 12345:	ok.
  4 < 40000:	ok.
   = :	ok.
  1 > :	ok.
  0 < 1:	ok.
  0.2 > 0.15:	ok.
  v0.2 < v0.15:	ok.
  0.200 > 0.15:	ok.
  v0.200 > v0.15:	ok.
  v0.200 < v0.15000:	ok.
  abc = Def:	ok.
  vabc = vDef:	ok.
  1.2.3.4 = 1.2.3.4:	ok.
  1.2.3.4 < 1.2.3.5:	ok.
  1.2.3.4 < 1.2.4.4:	ok.
  1.2.3.4 < 1.3.3.4:	ok.
  1.2.3.4 < 2.2.3.4:	ok.
  foo.com = foo.com:	ok.
  foo.com > bar.com:	ok.
  foobar.com > bar.com:	ok.
  barfoo.com > bar.com:	ok.
  foo.bar.com > bar.bar.com:	ok.
  foo.bar.x.com > bar.x.com:	ok.
  bar.foo.x.com > foo.x.com:	ok.
  foo.x.com > foo.bar.x.com:	ok.
